### ISSUE TITLE
Can't re-join a meeting if the previously associated server node is offline

### ISSUE THREAD

**Issue by lakano** [blindsidenetworks/scalelite#993]
**Describe the bug**
Hi!
Thank you for this amazing project!
Generally we create a room few days / hours before the real live meeting, and if we go inside the meeting when we prepare it, this room seems to be definitively assigned to the selected server. 

The problem is, if at the real time where the meeting should start, this assigned server is offline, Scalelite *should* assign a new server node.

**To Reproduce**
- Create a Room
- Start the meeting, and exit/stop the meeting 
- Shutdown the node server
- Wait 1 minute ( permit to SL to mark this server node offline ) 
- Start your meeting => Can't reach the initial assigned node server

***Deployment:***
1. Deployment: docker-compose 
2. Versions: 
blindsidenetwks/scalelite   v1-focal260-alpine-poller               36f60b9b9cb6   7 days ago     1.4GB
blindsidenetwks/scalelite   v1-focal260-alpine-recording-importer   fe78ddf1b746   7 days ago     1.4GB
blindsidenetwks/scalelite   v1-focal260-alpine-api                  04e1e94dd835   7 days ago     1.4GB
blindsidenetwks/scalelite   v1-focal260-alpine-nginx                06a7d986d2ad   7 days ago     28.6MB

**Comment by Ithanil** [comment]
Yes, this whole system is very broken currently. Meetings will also be assigned to disabled servers. Meetings will also remain assigned to removed servers. Thanks for creating the issue, somehow I didn't manage to make myself do it.

**Comment by Ithanil** [comment]
@farhatahmad The problem is that 1) the panicking doesn't actually disassociate the meetings from the server in the Redis and b) currently - at least since MT update and with MT enabled - meetings will be placed on disabled hosts.

Try the following: Have a server online, disable it and start a meeting few seconds later. It will (maybe) be placed on the disabled host and remain on it indefinitely, leading to the following XML error:

<response>
<returncode>FAILED</returncode>
<messageKey>serverUnavailable</messageKey>
<message>The server for this meeting is disabled/offline.</message>
</response>

**Comment by farhatahmad** [comment]
Yep locally atleast I can't reproduce it - will try it on one of our dev environments

**Comment by lakano** [comment]
Thank you for your interest on this problem!
Yes, I've wait more than 10 minutes.
May be SL keep this original server in Redis and don't want to assign to another server after ?
If you known the Redis keys, I can investigate to see the Redis keys settings.

**Comment by Ithanil** [comment]
> Yep locally atleast I can't reproduce it - will try it on one of our dev environments. To me - it sounds like there are issues with your Redis database causing what appears to be issues in Scalelite. I'll try to see if I can confirm that.
> 
> Can you reproduce it consistently?

Yes, absolutely. Meetings will always be placed on a disabled server if it is the first one in the list. That behavior is relatively new and probably happens only since we are using MT.
The issue with old meetings remaining associated with removed servers is not new however.

Please try the following script to check for inconsistencies in the Redis DB:

```
#!/bin/bash

#-------------------------------------------------
# Allow this script to terminate itself
#-------------------------------------------------
trap "exit 255" TERM;
export SCRIPT_PID=$$;
#-------------------------------------------------
function terminate() {
    printf "\n%s\n" "$*" >&2;
    kill -s TERM $SCRIPT_PID;
}

status=0
statustxt=

SERVERS1=$(diff -w <(redis-cli keys scalelite:server:* | cut -d: -f3|sort) <(redis-cli keys scalelite:*|egrep -v "server|meeting|health|protector_token|voice_bridges|tenant" |cut -d: -f2|sort) | egrep "[<>]" | tr '\n' ' ')

SERVERS=$(diff -w <(redis-cli keys scalelite:server:* | cut -d: -f3|sort) <(redis-cli smembers scalelite:servers |sort) | egrep "(<|>)" | tr '\n' ' ')

MEETINGS=$(diff -w <(redis-cli keys scalelite:meeting:* | cut -d: -f3|sort) <(redis-cli smembers scalelite:meetings |sort) | egrep "(<|>)" | tr '\n' ' ')

SERVENA=$(diff -w <(redis-cli smembers scalelite:server_enabled | cut -d: -f3|sort) <(redis-cli zrange scalelite:server_load 0 -1 |sort) | egrep "[<>]" | tr '\n' ' ')

MEETSERV=$(for meeting in $(redis-cli keys scalelite:meeting:* | cut -d\" -f3); do redis-cli smembers scalelite:servers | grep -q "$(redis-cli hget $meeting server_id)" || echo -n $meeting | cut -d: -f3 | tr '\n' ' ' ; done )

if [[ -n $SERVERS1 ]]
then
    redisscalelite="scalelite:server:* <-> scalelite:*:$SERVERS1 "
    statustxt="servers differ: "
    status=1
fi

if [[ -n $SERVERS ]]
then
    redisscalelite="$redisscalelite; scalelite:server:* <-> scalelite:servers:$SERVERS "
    statustxt="servers differ: "
    status=1
fi

if [[ -n $MEETINGS ]]
then
    redisscalelite="$redisscalelite; scalelite:meeting:* <-> scalelite:meetings:$MEETINGS "
    statustxt="$statustxt meetings differ: "
    status=1
fi

if [[ -n $SERVENA ]]
then
    redisscalelite="$redisscalelite; scalelite:server_enabled <-> scalelite:server_load:$SERVENA "
    statustxt="$statustxt enabled servers differ: "
    status=2
fi

if [[ -n $MEETSERV ]]
then
    redisscalelite="$redisscalelite; scalelite:server missing for meeting:$MEETSERV "
    statustxt="$statustxt server for meeting missing: "
    status=2
fi


#     status
#     |       item name
#     |       |               perfdata
#     |       |               | check output
#     |       |               | |
echo "$status redis-scalelite - $statustxt${redisscalelite:--}"

```

I've removed the code that is used to actually repair the issues, so this script does not change anything in the DB. A "good" output will look like `0 redis-scalelite - -` (this script is actually a Check for CheckMk, that's the reason for the format).

**Comment by JeanPluzo** [comment]
Hi,
I've been following this issue. I don't know if it can be of any help but here's some info:
- made a BBB server go offline without `rake cordon` (turned BBB "off" with `bbb-conf --stop`)
- waited for said server to be `STATUS offline` in `rake status` on SL
- started 2 different sessions through GL
- was sent to other online servers
- finished sessions
- brought offline server back online
- started yet another session through GL
- was sent to 1st server

Some notes:
- server which was online/offline/online is the 1st server to be chosen on the SL pool
- SL 1.3.4, no MT (of course)
- GL 2.14.8
- BBB 2.5.18

Summary
Yes, I know my setup is kind of outdated. We haven't been able to upgrade to 2.6. Because of the different bugs in recent SL version (sha, recordings, etc) we're hesitant to upgrade to a recent version (If it ain't broken don't fix it).
It seems SL without MT does what it's supposed to do. After the dial-in and MT updates in SL it seems to me/us there are still some bugs to be patched. We're waiting for SL to have fewer bugs to update to a recent version.
All in all, keep up the good/great work!!

Regards,
J.

**Comment by lakano** [comment]
I've done some more tests. 
Even if I manually rake servers:panic[id] the server, I can't go anymore in the room (others BBB servers are enabled/online)
Can you give us some details to what we need to remove in PostgreSQL/Redis to unattach a meeting to a server please? This could permit us to have at least a workaround until this bug is fixed please?

**Comment by lakano** [comment]
If I delete the key in redis ( similar to « scalelite:meeting:xxxxxxxxxxxxx » ) this permit me to go again in the room in another available BBB server.
But it's not practical, first we need to detect all meetings ID associated to the server, then to delete them from Redis.
It's not the goal of a "panic" query ? (and Poll also run a panic when a server turn offline)

**Comment by Ithanil** [comment]
@lakano Actually I removed the "repair" functionality out of the script I posted above, to make it clearer. Below is a version which accepts "--repair" as argument... it will clean both removed servers and meetings on the removed servers (execute it it repeatedly until output is `0 redis-scalelite - -`). But it doesn't care for disabled (but not removed) servers and their meetings.

```
#!/bin/bash

REPAIR=0
if [[ $1 == "--repair" ]]
then
    REPAIR=1
fi

status=0
statustxt=

SERVERS1=$(diff -w <(redis-cli keys scalelite:server:* | cut -d: -f3|sort) <(redis-cli keys scalelite:*|egrep -v "server|meeting|health|protector_token|voice_bridges|tenant" |cut -d: -f2|sort) | egrep "[<>]" | tr '\n' ' ')

SERVERS=$(diff -w <(redis-cli keys scalelite:server:* | cut -d: -f3|sort) <(redis-cli smembers scalelite:servers |sort) | egrep "(<|>)" | tr '\n' ' ')

MEETINGS=$(diff -w <(redis-cli keys scalelite:meeting:* | cut -d: -f3|sort) <(redis-cli smembers scalelite:meetings |sort) | egrep "(<|>)" | tr '\n' ' ')

SERVENA=$(diff -w <(redis-cli smembers scalelite:server_enabled | cut -d: -f3|sort) <(redis-cli zrange scalelite:server_load 0 -1 |sort) | egrep "[<>]" | tr '\n' ' ')

MEETSERV=$(for meeting in $(redis-cli keys scalelite:meeting:* | cut -d\" -f3); do redis-cli smembers scalelite:servers | grep -q "$(redis-cli hget $meeting server_id)" || echo -n $meeting | cut -d: -f3 | tr '\n' ' ' ; done )

if [[ -n $SERVERS1 ]]
then
    redisscalelite="scalelite:server:* <-> scalelite:*:$SERVERS1 "
    statustxt="servers differ: "
    status=1
    if [[ $REPAIR -eq 1 ]]
    then
        for server in $(echo $SERVERS1 | tr '>' ' ')
        do
            echo $server
            redis-cli del scalelite:$server
        done
    fi
fi

if [[ -n $SERVERS ]]
then
    redisscalelite="$redisscalelite; scalelite:server:* <-> scalelite:servers:$SERVERS "
    statustxt="servers differ: "
    status=1
    if [[ $REPAIR -eq 1 ]]
    then
	redis-cli srem scalelite:servers $(echo $SERVERS | tr '>' ' ')
    fi
fi

if [[ -n $MEETINGS ]]
then
    redisscalelite="$redisscalelite; scalelite:meeting:* <-> scalelite:meetings:$MEETINGS "
    statustxt="$statustxt meetings differ: "
    if [[ $REPAIR -eq 1 ]]
    then
	redis-cli srem scalelite:meetings $(echo $MEETINGS | tr '>' ' ')
    fi
    status=1
fi

if [[ -n $SERVENA ]]
then
    redisscalelite="$redisscalelite; scalelite:server_enabled <-> scalelite:server_load:$SERVENA "
    statustxt="$statustxt enabled servers differ: "
    status=2
fi

if [[ -n $MEETSERV ]]
then
    redisscalelite="$redisscalelite; scalelite:server missing for meeting:$MEETSERV "
    statustxt="$statustxt server for meeting missing: "
    status=2
    if [[ $REPAIR -eq 1 ]]
    then
        for meeting in $MEETSERV
        do
	    echo $meeting
	    redis-cli del scalelite:meeting:$meeting
	done
    fi
fi

echo "$status redis-scalelite - $statustxt${redisscalelite:--}"

```

**Comment by Ithanil** [comment]
@farhatahmad Were you still not able to reproduce any issues, even with MT enabled?

**Comment by Ithanil** [comment]
I've found one of the problems, it's in panic API call and rake panic tasks. I'll open a separate issue & PR (just as proposal for a fix).

**Comment by Ithanil** [comment]
@lakano Can you confirm that the issue is fixed in [v1.5.1.6](https://github.com/blindsidenetworks/scalelite/releases/tag/v1.5.1.6)?

**Comment by lakano** [comment]
I'm really sorry, I can't tell because I didn't keep the tested servers, but from what I've read on your patch, this is surely the reason of my reported issue. So I'm closing this one and if someone get again this problem, he could open a new issue.
